README.md
[![Unix Build Status][ci-img]][ci-url]
[![Windows Build Status][ci-win-img]][ci-win-url]
[![Code Climate][clim-img]][clim-url]

# haraka-plugin-batv-etcd-config

This Bounce Address Tag Validation plugin uses "etcd" to access to its config variables. Otherwise, it functions exactly the same way as the batv plugin: https://github.com/PartyPancakess/haraka-plugin-batv


This plugin uses srs.js script directly. For more information, please check: https://www.npmjs.com/package/srs.js

& IMPORTANT: this plugin must appear in  `config/plugins`  before other plugins that run on hook_rcpt

## How it works
Assume that the user uses the address example@domain.com and will send an e-mail.

Before the e-mail is sent, example@domain.com will automatically change to:
SRS0=HHH=TT=domain.com=example@domain.com

If the e-mail bounces, after checking if the key is correct or not or whether there is a key at all, it will be forwarded to example@domain.com.


## Configuration
The running etcd server must have 2 variables exclusively for this plugin: config_batv_secret and config_batv_maxAge. 
The value of "config_batv_secret" must be a string, which is the secret key of the srs.
The value of "config_batv_maxAge" must contain a string ("day" or "second"), which determines the type of the maximum age, and an integer, separated by a hyphen.
Below can be found example cases:
```
config_batv_secret: "asecretkey"
config_batv_maxAge: "day-21"

or

config_batv_secret: "asecretkey"
config_batv_maxAge: "second-1814400"
```


## Example etcd Configuration
```
etcdctl put config_batv_maxAge day-7
etcdctl put config_batv_secret secretkey
```





<!-- leave these buried at the bottom of the document -->
[ci-img]: https://github.com/haraka/haraka-plugin-batv-etcd-config/workflows/Plugin%20Tests/badge.svg
[ci-url]: https://github.com/haraka/haraka-plugin-batv-etcd-config/actions?query=workflow%3A%22Plugin+Tests%22
[ci-win-img]: https://github.com/haraka/haraka-plugin-batv-etcd-config/workflows/Plugin%20Tests%20-%20Windows/badge.svg
[ci-win-url]: https://github.com/haraka/haraka-plugin-batv-etcd-config/actions?query=workflow%3A%22Plugin+Tests+-+Windows%22
[clim-img]: https://codeclimate.com/github/haraka/haraka-plugin-batv-etcd-config/badges/gpa.svg
[clim-url]: https://codeclimate.com/github/haraka/haraka-plugin-batv-etcd-config
[npm-img]: https://nodei.co/npm/haraka-plugin-batv-etcd-config.png
[npm-url]: https://www.npmjs.com/package/haraka-plugin-batv-etcd-config


Changes.md
## 1.0.0 - 2020-12-13

- Initial release


test/index.js

// node.js built-in modules
const assert   = require('assert')

// npm modules
const fixtures = require('haraka-test-fixtures')

// start of tests
//    assert: https://nodejs.org/api/assert.html
//    mocha: http://mochajs.org

beforeEach(function (done) {
  this.plugin = new fixtures.plugin('batv-etcd-config')
  done()  // if a test hangs, assure you called done()
})

describe('batv-etcd-config', function () {
  it('loads', function (done) {
    assert.ok(this.plugin)
    done()
  })
})

describe('load_batv-etcd-config_ini', function () {
  it('loads batv-etcd-config.ini from config/batv-etcd-config.ini', function (done) {
    this.plugin.load_batv-etcd-config_ini()
    assert.ok(this.plugin.cfg)
    done()
  })

  it('initializes enabled boolean', function (done) {
    this.plugin.load_batv-etcd-config_ini()
    assert.equal(this.plugin.cfg.main.enabled, true, this.plugin.cfg)
    done()
  })
})

describe('uses text fixtures', function () {
  it('sets up a connection', function (done) {
    this.connection = fixtures.connection.createConnection({})
    assert.ok(this.connection.server)
    done()
  })

  it('sets up a transaction', function (done) {
    this.connection = fixtures.connection.createConnection({})
    this.connection.transaction = fixtures.transaction.createTransaction({})
    // console.log(this.connection.transaction)
    assert.ok(this.connection.transaction.header)
    done()
  })
})

srs.js
"use strict";

var crypto = require("crypto");

// This timestamp code is based on the C libsrs2 implementation.
var timeBaseChars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567",
  timeSize = 2,
  timeBaseBits = 5,
  timePrecision = 1, // second based
  timeSlots = (1<<(timeBaseBits<<(timeSize-1))),
  validSeparators = "=-+";

function makeTimestamp() {
  var now = Math.round(Date.now() / 1000 / timePrecision);
  var str = timeBaseChars[now & ((1 << timeBaseBits) - 1)];
  now = now >> timeBaseBits;
  str = timeBaseChars[now & ((1 << timeBaseBits) - 1)] + str;
  return str;
}

function createHash(secret, timestamp, domain, local) {
  var hmac = crypto.createHmac("sha1", secret);
  Array.prototype.slice.call(arguments, 1).forEach(function(data) {
    hmac.update(data);
  });
  return hmac.digest("hex").substring(0, 4);
}

function checkTimestamp(str, maxAge) {
  var then = 0;
  for (var i = 0; i < str.length; i++) {
    then = (then << timeBaseBits) | timeBaseChars.indexOf(str[i].toUpperCase());
  }

  var now = Math.round(Date.now() / 1000 / timePrecision) % timeSlots;
  while(now < then) {
    now = now + timeSlots;
  }
  return now <= then + maxAge;
}

function SRS(options) {
  options = options || {};
  this.secret = options.secret;
  this.separator = (options.separator || "=")[0];
  this.maxAge = options.maxAge || (21 * 24 * 60 * 60); // 21 days

  if (!this.secret) {
    throw new TypeError("A secret must be provided");
  }

  if (validSeparators.indexOf(this.separator) < 0) {
    throw new TypeError("Invalid separator");
  }

  this.srs0Re = new RegExp("SRS0[\\-=+]([0-9a-f]{4})=" + 
                            "([" + timeBaseChars + "]{2})=" +
                            "([^=]*)=(.*)");

  this.srs1Re = new RegExp("SRS1[\\-=+]([0-9a-f]{4})=([^=]*)=(.*)");
}

SRS.prototype.isSrs0 = function(local) {
  return local.indexOf("SRS0") === 0;
};

SRS.prototype.isSrs1 = function(local) {
  return local.indexOf("SRS1") === 0;
};

SRS.prototype.rewrite = function(local, domain) {
  if (this.isSrs0(local)) {
    // Create a guarded address.
    var guarded = local.substring(4);
    return "SRS1" + this.separator + createHash(this.secret, domain, guarded) +
           this.separator + domain + this.separator + guarded;
  } else if (this.isSrs1(local)) {
    var match = this.srs1Re.exec(local);
    if (!match) {
      throw new Error("Attempted to rewrite invalid SRS1 address.");
    }
    return "SRS1" + this.separator + createHash(this.secret, match[2], match[3]) +
           "=" + match[2] + "=" + match[3];
  }

  var timestamp = makeTimestamp();
  var hash = createHash(this.secret, timestamp, domain, local);    

  return "SRS0" + this.separator + hash + "=" + timestamp +
                  "=" + domain + "=" + local;
};

SRS.prototype.reverse = function(address, addressDomain) {
  var matches, hash, timestamp, domain, local, expectedHash;

  if (this.isSrs0(address)) {
    matches = this.srs0Re.exec(address);
    if (!matches) {
      throw new TypeError("Unrecognized SRS0 format");
    }
    
    hash = matches[1];
    timestamp = matches[2];
    domain = matches[3];
    local = matches[4];

    expectedHash = createHash(this.secret, timestamp, domain, local);
    if (expectedHash !== hash) {
      throw new TypeError("Invalid signature");
    }

    if (!checkTimestamp(timestamp, this.maxAge)) {
      throw new TypeError("Address has expired");
    }

    return [local, domain];
  } else if (this.isSrs1(address)) {
    matches = this.srs1Re.exec(address);
    if (!matches) {
      throw new TypeError("Unrecognized SRS1 format");
    }
    
    hash = matches[1];
    domain = matches[2];
    local = matches[3];

    expectedHash = createHash(this.secret, domain, local);
    if (expectedHash !== hash) {
      throw new TypeError("Invalid signature");
    }

    return ["SRS0" + local, domain];
  }

  return null;
};

module.exports = SRS;


index.js
// Put this plugin's name before all the rcpt_to plugins in the config/plugins file.
// This plugin requires haraka-necessary-helper-plugins to be in the same folder with it: https://github.com/PartyPancakess/haraka-necessary-helper-plugins
'use strict'

const Address = require('../haraka-necessary-helper-plugins/address-rfc2821').Address;

var SRS = require("./srs.js");
var rewriter;

const { Etcd3 } = require('../haraka-necessary-helper-plugins/etcd3');
const client = new Etcd3();

exports.register = function () {
  this.inherits('queue/discard');

  this.load_batv_ini();
  
  this.register_hook('data_post', 'relay');
  this.register_hook('rcpt', 'rcpt');
}

exports.load_batv_ini = function () {
  const plugin = this;

  const tempConfig = {
    srs: {
        secret: 'asecretkey',
        maxAgeSeconds: '-1',
        maxAgeDays: '-1'
    }
  }
  plugin.cfg = tempConfig;

  (async () => {
    plugin.cfg.srs.secret = await client.get('config_batv_secret').string();
    const age = await (await client.get('config_batv_maxAge').string()).split("-");

    if(age[0]==="day") {
      plugin.cfg.srs.maxAgeDays = age[1];
      plugin.cfg.srs.maxAgeSeconds = '-1';
    }
    else if(age[0]==="second") {
      plugin.cfg.srs.maxAgeSeconds = age[1];
      plugin.cfg.srs.maxAgeDays = '-1';
    }
    this.createSrs(plugin);
  })();

  client.watch()
  .key('config_batv_secret')
  .create()
  .then(watcher => {
    watcher
      .on('disconnected', () => console.log('disconnected...'))
      .on('connected', () => console.log('successfully reconnected!'))
      .on('put', res => {
        plugin.cfg.srs.secret = res.value.toString();
        console.log('config_batv_secret got set to:', res.value.toString());
        this.createSrs(plugin);
      });
  });

  client.watch()
  .key('config_batv_maxAge')
  .create()
  .then(watcher => {
    watcher
      .on('disconnected', () => console.log('disconnected...'))
      .on('connected', () => console.log('successfully reconnected!'))
      .on('put', res => {
        const age = res.value.toString().split("-");
        if(age[0]==="day") {
          plugin.cfg.srs.maxAgeDays = age[1];
          plugin.cfg.srs.maxAgeSeconds = '-1';
        }
        else if(age[0]==="second") {
          plugin.cfg.srs.maxAgeSeconds = age[1];
          plugin.cfg.srs.maxAgeDays = '-1';
        }
        console.log('config_batv_maxAge got set to:', res.value.toString());
        console.log(plugin.cfg);
        // plugin.load_batv_ini();
        this.createSrs(plugin);
      });
  });

  this.createSrs(plugin);
}

exports.rcpt = function (next, connection, params) { // Check the rcpt and decide if it is spam or not.
  var txn = connection.transaction;
  const plugin = this;

  if(!connection.relaying && txn.mail_from.isNull()) { // Incoming
    var oldUser = txn.rcpt_to[0].user;
    var reversed = rewriter.reverse(oldUser);
    
    if(reversed === null || reversed === undefined) {
      connection.logdebug(plugin, "marking " + txn.rcpt_to + " for drop");
      connection.transaction.notes.discard = true;
    }
    else {
      txn.rcpt_to.pop();
      var sendTo = reversed[0] + "@" + reversed[1];
      txn.rcpt_to.push(new Address(`<${sendTo}>`));
    }
  }

  next();
}

exports.relay = function (next, connection) { // Change the sender address of outgoing e-mails.
  var txn = connection.transaction;
  if (connection.relaying) {    
    var oldMailFrom = txn.mail_from;
    var rewritten = rewriter.rewrite(oldMailFrom.user, oldMailFrom.host);
    var newMail = new Address(`<${rewritten}@${oldMailFrom.host}>`);

    txn.mail_from = newMail;
  }

  next();
}

exports.createSrs = function (plugin) {
  var maxAgeSeconds = 21 * 24 * 60 * 60; // default: 21 days 
  if(plugin.cfg.srs.maxAgeDays !== '-1') maxAgeSeconds = plugin.cfg.srs.maxAgeDays * 24 * 60 * 60;
  else if(plugin.cfg.srs.maxAgeSeconds !== '-1') maxAgeSeconds = plugin.cfg.srs.maxAgeSeconds;

  rewriter = new SRS({
    secret: plugin.cfg.srs.secret,
    maxAge: maxAgeSeconds
  });
}
